File: problem-solving/_lc-0115-hard-distinct-subsequences/1-top-down.py
"""
https://leetcode.com/problems/distinct-subsequences/

Complexity
----------
Time: O(m*n)
Space: O(m*n)
"""

class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        cache = {}
        def _numDistinct(i, j):
            if j == len(t):
                return 1
            if i == len(s):
                return 0
            if (i, j) in cache:
                return cache[(i, j)]
            if s[i] == t[j]:
                cache[(i, j)] = _numDistinct(i + 1, j) + _numDistinct(i + 1, j + 1)
            else:
                cache[(i, j)] = _numDistinct(i + 1, j)
            return cache[(i, j)]
        return _numDistinct(0, 0)

File: problem-solving/_lc-0115-hard-distinct-subsequences/1-top-down.go
/*
https://leetcode.com/problems/distinct-subsequences/

Complexity
----------
Time: O(m*n)
Space: O(m*n)
*/

func numDistinct(s string, t string) int {
    cache := make(map[string]int)
    var F func(int, int) int
    F = func(i, j int) int {
        if j == len(t) {
            return 1
        }
        if i == len(s) {
            return 0
        }
        key := fmt.Sprintf("%d-%d", i, j)
        if v, ok := cache[key]; ok {
            return v
        }
        if s[i] == t[j] {
            cache[key] = F(i + 1, j) + F(i + 1, j + 1)
        } else {
            cache[key] = F(i + 1, j)
        }
        return cache[key]
    }
    return F(0, 0)
}
